File: android-frontend/SmartMedicine/app/src/main/java/com/czy/smartmedicine/viewModel/base/BaseVmActivity.java
package com.czy.smartmedicine.viewModel.base;


import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.viewbinding.ViewBinding;

import com.czy.baseutil.activity.BaseActivity;
import com.czy.smartmedicine.MainApplication;


/**
 * 解决ViewBinding重复代码    （通过反射实现）
 * @param <VB>     视图绑定类型
 * @param <VM>     视图模型类型
 */
public abstract class BaseVmActivity<VB extends ViewBinding, VM extends ViewModel> extends BaseActivity<VB> {

    protected VB binding;
    protected VM vm;
    private final Class<VM> viewModelClass;

    public abstract VB getBinding();

    public BaseVmActivity(Class<? extends FragmentActivity> classType, Class<VM> viewModelClass){
        super(classType);
        this.viewModelClass = viewModelClass;
    }

    protected void initViewModel() {
        ApiViewModelFactory apiViewModelFactory = new ApiViewModelFactory(
                MainApplication.getApiRequestImplInstance(),
                MainApplication.getInstance().getMessageSender()
        );

        // 使用 ViewModelProvider 创建 ViewModel 实例
        vm = new ViewModelProvider(this, apiViewModelFactory).get(viewModelClass);
    }
}

File: android-frontend/SmartMedicine/app/src/main/java/com/czy/smartmedicine/utils/BaseVmActivity.kt
package com.czy.smartmedicine.utils

import android.os.Bundle
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.ViewModel
import androidx.viewbinding.ViewBinding
import com.czy.baseutil.viewModel.ViewModelUtil
import com.czy.smartmedicine.MainApplication
import com.czy.smartmedicine.viewModel.base.ApiViewModelFactory
import kotlin.reflect.KClass

abstract class BaseVmActivity<VB : ViewBinding, VM : ViewModel> (
    activityClassType: KClass<out FragmentActivity>,
    private val vmClassType: KClass<VM>
) : FragmentActivity(){

    protected open lateinit var vm: VM
    protected open lateinit var binding: VB

    private val activityName : String = activityClassType.java.name
    protected open val TAG : String = activityName

    abstract fun initBinding(): VB

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.binding = initBinding()
        setContentView(binding.root)

        initViewModel()

        initView()

        setListener()
    }

    protected open fun initView(){

    }

    protected open fun setListener() {

    }

    protected open fun initViewModel() {
        val apiViewModelFactory = ApiViewModelFactory(
            MainApplication.getApiRequestImplInstance(),
            MainApplication.getInstance().getMessageSender()
        )

        vm = ViewModelUtil.newViewModel(this, apiViewModelFactory, vmClassType.java)
    }
}